src/kernels/MIOpenGroupNorm.cpp

```diff
@@ -0,0 +1,129 @@
+#define CVT_FP32_2ACCUM(x) (x)
+#endif
+
+extern "C" __global__ void GroupNormFwdContiguous(const FLOAT* __restrict__ x,
+                                                  FLOAT* __restrict__ y,
+                                                  const FLOAT* __restrict__ weight,
+                                                  const FLOAT* __restrict__ bias,
+                                                  FLOAT_ACCUM* __restrict__ mean,
+                                                  FLOAT_ACCUM* __restrict__ rstd,
+                                                  float eps,
+                                                  uint64_t num_groups,
+                                                  uint64_t num_channels,
+                                                  uint64_t numel_per_channel,
+                                                  bool mode)
+{
+    /*
+     * Each group works on a single channel.
+     * Example)
+     * x dim = {N, C, L}, normalized shape = {C, L}
+     * outer_size = N, inner_size = C * L
+     *
+     * Example2)
+     * x dim = {N, C, L}, normalized shape = {L}
+     * outer_size = N * C, inner_size = L
+     *
+     * => gws = {outer_size * LOCAL_SIZE}, lws = {LOCAL_SIZE}
+     */
+
+    /*
+     * Reduction to calculate mean and rstd
+     */
+
+    const uint64_t gid = blockIdx.x;
+    const uint64_t lid = threadIdx.x;
+
+    const size_t inner_size = numel_per_channel * num_channels / num_groups;
+
+    FLOAT_ACCUM pmean = static_cast<FLOAT_ACCUM>(0);
+    FLOAT_ACCUM pvar  = static_cast<FLOAT_ACCUM>(0);
+    __shared__ FLOAT_ACCUM ltmp1[LOCAL_SIZE];
+    __shared__ FLOAT_ACCUM ltmp2[LOCAL_SIZE];
+
+    // reduce sum for mean and var
+    for(uint64_t i = lid; i < inner_size; i += LOCAL_SIZE)
+    {
+        size_t x_idx = gid * inner_size + i;
+
+        FLOAT_ACCUM tmp = CVT_FLOAT2ACCUM(x[x_idx]);
+        pmean += tmp;
+        pvar += tmp * tmp;
+    }
+
+    ltmp1[lid] = pmean;
+    ltmp2[lid] = pvar;
+    __syncthreads();
+    for(uint32_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            ltmp1[lid] += ltmp1[lid + i];
+            ltmp2[lid] += ltmp2[lid + i];
+        }
+        __syncthreads();
+    }
+    pmean             = ltmp1[0] / inner_size;
+    pvar              = ltmp2[0] / inner_size - pmean * pmean;
+    FLOAT_ACCUM prstd = rsqrt(pvar + FLOAT_ACCUM(eps));
+
+    if(lid == 0)
+    {
+        if(mean)
+            mean[gid] = pmean;
+        if(rstd)
+            rstd[gid] = prstd;
+    }
+
+    // forward calculation
+    for(uint64_t i = lid; i < inner_size; i += LOCAL_SIZE)
+    {
+        size_t idx = gid * inner_size + i;
+
+        FLOAT_ACCUM pweight;
+        FLOAT_ACCUM pbias;
+
+        size_t c = (idx / numel_per_channel) % num_channels;
+        pweight  = mode ? CVT_FLOAT2ACCUM(weight[c]) : CVT_FP32_2ACCUM(1.0f);
+        pbias    = mode ? CVT_FLOAT2ACCUM(bias[c]) : static_cast<FLOAT>(0);
```

Comment:
There is no need to calculate `c` and cast should be done in a uniform way.
```suggestion
        size_t c = mode ? (idx / numel_per_channel) % num_channels : 0;

        FLOAT_ACCUM pweight  = mode ? CVT_FLOAT2ACCUM(weight[c]) : static_cast<FLOAT>(1);
        FLOAT_ACCUM pbias    = mode ? CVT_FLOAT2ACCUM(bias[c]) : static_cast<FLOAT>(0);
```

Reply:
Fixed as per your comment.